core/services/pipeline/orm.go

```diff
@@ -159,6 +157,47 @@ func (o *orm) CreateRun(ctx context.Context, jobID int32, meta map[string]interf
             SELECT ? AS pipeline_run_id, id AS pipeline_task_spec_id, type, index, NOW() AS created_at
             FROM pipeline_task_specs
             WHERE pipeline_spec_id = ?`, run.ID, run.PipelineSpecID).Error
+		if err != nil {
+			return err
+		}
+
+		// Ammend the TaskRuns with their predecessors.
+		var tr []TaskRun
+		err = tx.Find(&tr).Where("pipeline_run_id = ?", runID).Error
+		if err != nil {
+			return err
+		}
+
+		var ts []TaskSpec
+		err = tx.Find(&ts).Where("pipeline_spec_id = ?", run.PipelineSpecID).Error
+		if err != nil {
+			return err
+		}
+
+		var preds = make(map[int32][]int32)
+		for _, taskSpec := range ts {
+			if taskSpec.SuccessorID.Valid {
+				preds[int32(taskSpec.SuccessorID.Int64)] = append(preds[int32(taskSpec.SuccessorID.Int64)], taskSpec.ID)
+			}
+		}
+		for _, taskRun := range tr {
+			if preds, ok := preds[taskRun.PipelineTaskSpecID]; ok {
+				// Find the corresponding taskRun for each pred
+				// TODO: clever optimize?
```

Comment:
todo spend a little more time on this - its sort of like a graph cloning problem

Reply:
Can we pull this out into a neatly named function?

Reply:
Also I have a working optimised path for doing something like this, you might be able to use it:

https://github.com/smartcontractkit/chainlink/commit/7a85aeeb11c45c2d583caa281177c6eedf22486d#diff-93a69845888e9fa37d8e5ec7d17ff8f7003b52f598cf0367b1944ed1466966eaR205

Check the whole diff here, there might also be some tests you can use:

https://github.com/smartcontractkit/chainlink/compare/chore/176526416-optimize-work-queue?expand=1

Reply:
I think they are slightly different problems, but it was fairly simple to optimize so I just optimized it and pulled out a helper/unit test.